### PR TITLE
Add mirror to Gluon

### PR DESCRIPTION
**Pull request by piiswrong**
New hybridize option use_mirror for trading speed for memory

### CODE REVIEW

File: src/imperative/cached_op.cc [hunk 1]
```diff
@@ -89,6 +89,22 @@ struct CachedOp::CachedOpState {
   std::multimap<size_t, NDArray> bwd_reuse_pool;
 };
 
+template<typename ValueType>
+inline ValueType get_node_attr(
+    const nnvm::Node& node,
+    const std::string& key, ValueType default_value) {
```
Comment (szha) [inline src/imperative/cached_op.cc]: this looks generic enough to be in a utils file.

File: python/mxnet/gluon/_memmonger.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+# under the License.
+
+# coding: utf-8
+# pylint: disable=
+
+import math
+
+from .. import cpu
+
+def prod(shape):
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/_memmonger.py]: make private _prod ?

File: python/mxnet/gluon/_memmonger.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    sym : symbol
+        Input configuration of symbols.
+        The user need to pre-mark the attribute "mirror_stage" on the nodes
+        that can be book-kept as stage
+        The algorithm will decide whether to disbale mirror on the stage nodes.
+    threshold: integer
+        A tuning parameter to tune the approximate size of each stage blocks
+    plan_info: dict, optional
+        Used to hold plan information.
+    **kwargs:
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/_memmonger.py]: is that supposed to be `inputs` and `params`?

File: python/mxnet/gluon/_memmonger.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    if plan_info is not None:
+        plan_info['max_size'] = max_size
+        plan_info['save_size'] = save_size
+    return sym
+
+
+def get_cost(sym, inputs, params):
+    """Get the cost of the current symbolic plan by running bind on CPU.
+    sym : Symbolic Variable
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/_memmonger.py]: `input` and `params` doc string?

File: python/mxnet/gluon/_memmonger.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+
+def search_plan(sym, inputs, params, ntrials=6):
+    """Quickly heurestic search over possible plans to find good memory plan.
+
+    Parameters
+    ----------
+    sym : symbolic
+       Symbolic configurations
+    ntrials: integer
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/_memmonger.py]: default value

File: python/mxnet/gluon/_memmonger.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+    texec = sym.simple_bind(ctx=cpu(),
+                            grad_req=grad_reqs,
+                            type_dict=type_dict,
+                            **shape_dict)
+    return int(texec.debug_str().split('\n')[-3].split()[1])
+
+
+def search_plan(sym, inputs, params, ntrials=6):
+    """Quickly heurestic search over possible plans to find good memory plan.
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/_memmonger.py]: Could you describe the heuristic* used to find the memory plan? Could you also describe for the users according to which metric do we consider a memory plan good?

File: python/mxnet/gluon/_memmonger.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        sym = make_mirror_plan(sym, inputs, params, threshold, info)
+        cost = get_cost(sym, inputs, params)
+        save_size = info['save_size'] >> 20
+        local_size = info['max_size'] >> 20
+        guess = int(math.sqrt(save_size * local_size / 2))
+        if min_cost is None or min_cost > cost:
+            min_cost = cost
+        if min_threshold is None or local_size < min_threshold:
+            min_threshold = local_size
+        print ("Search threshold=%d MB, cost=%d MB" % (threshold, cost))
```
Comment (ThomasDelteil) [inline python/mxnet/gluon/_memmonger.py]: can we use the `logging` API rather than print here?

**Pull request closed without merging**